File: databench/service/routers/datasets.py
"""Dataset endpoints: ingest, inspect, preview, export."""

from __future__ import annotations

import json
import os
import tempfile
from itertools import islice
from typing import Optional

from fastapi import APIRouter, Depends, File, Query, UploadFile
from fastapi.responses import StreamingResponse

from databench import Manifest, Workspace
from databench.schema import Kind
from databench.workspace import _export_record

from ..deps import get_workspace
from ..schemas import IngestSamplesRequest, SamplesPage

router = APIRouter(tags=["datasets"])


@router.post("/datasets", response_model=Manifest)
def ingest_samples(req: IngestSamplesRequest, ws: Workspace = Depends(get_workspace)) -> Manifest:
    """Ingest a list of typed samples (JSON body) as a new dataset."""

    ds = ws.add_samples(req.samples, name=req.name, message=req.message)
    return ds.manifest


@router.post("/datasets:ingest-jsonl", response_model=Manifest)
async def ingest_jsonl(
    file: UploadFile = File(...),
    name: Optional[str] = Query(None),
    kind: Optional[Kind] = Query(None),
    source: Optional[str] = Query(None),
    ws: Workspace = Depends(get_workspace),
) -> Manifest:
    """Ingest an uploaded ``.jsonl`` file (kind auto-detected per line)."""

    payload = await file.read()
    tmp = tempfile.NamedTemporaryFile("wb", suffix=".jsonl", delete=False)
    try:
        tmp.write(payload)
        tmp.close()
        ds = ws.add_jsonl(tmp.name, name=name, kind=kind, source=source)
    finally:
        os.unlink(tmp.name)
    return ds.manifest


@router.get("/datasets/{ref}", response_model=Manifest)
def get_dataset(ref: str, ws: Workspace = Depends(get_workspace)) -> Manifest:
    """Return the manifest (version, name, row count, kind histogram) of a dataset."""

    return ws.get(ref).manifest


@router.get("/datasets/{ref}/samples", response_model=SamplesPage)
def preview_samples(
    ref: str,
    limit: int = Query(20, ge=1, le=500),
    offset: int = Query(0, ge=0),
    ws: Workspace = Depends(get_workspace),
) -> SamplesPage:
    """Paginated preview of a dataset's samples."""

    ds = ws.get(ref)
    items = list(islice(ds.to_samples(), offset, offset + limit))
    return SamplesPage(total=len(ds), limit=limit, offset=offset, items=items)


@router.get("/datasets/{ref}/export")
def export_dataset(
    ref: str,
    fmt: str = Query("messages-jsonl"),
    ws: Workspace = Depends(get_workspace),
) -> StreamingResponse:
    """Stream a dataset as training-ready JSONL."""

    ds = ws.get(ref)

    def lines():
        for sample in ds.to_samples():
            yield json.dumps(_export_record(sample, fmt), ensure_ascii=False) + "\n"

    filename = f"{ds.name or ds.version[:12]}.jsonl"
    return StreamingResponse(
        lines(),
        media_type="application/x-ndjson",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )

File: apps/api/src/routes/datasets.ts
import { randomUUID } from 'node:crypto'
import { rm, writeFile } from 'node:fs/promises'
import { tmpdir } from 'node:os'
import { join, parse } from 'node:path'
import {
  BadInputError,
  DatasetResponseSchema,
  ExportDatasetQuerySchema,
  IngestJsonlQuerySchema,
  IngestSamplesOpenApiRequestSchema,
  IngestSamplesRequestSchema,
  PaginationQuerySchema,
  SamplesPageOpenApiSchema,
  toJsonCompatible,
} from '@databench/schema'
import { createRoute, type OpenAPIHono } from '@hono/zod-openapi'
import { z } from 'zod'
import type { ApiEnv } from '../context.js'
import { getWorkspace } from '../context.js'
import { DEFAULT_ERROR_RESPONSES, jsonResponse } from '../openapi.js'

const ingestSamplesRoute = createRoute({
  method: 'post',
  path: '/v1/datasets',
  request: {
    body: {
      content: {
        'application/json': {
          schema: IngestSamplesOpenApiRequestSchema,
        },
      },
    },
  },
  responses: {
    200: jsonResponse(DatasetResponseSchema, 'Dataset manifest'),
    ...DEFAULT_ERROR_RESPONSES,
  },
})

const ingestJsonlRoute = createRoute({
  method: 'post',
  path: '/v1/datasets:ingest-jsonl',
  request: {
    query: IngestJsonlQuerySchema,
    body: {
      content: {
        'multipart/form-data': {
          schema: z.object({
            file: z.any(),
          }),
        },
      },
    },
  },
  responses: {
    200: jsonResponse(DatasetResponseSchema, 'Dataset manifest'),
    ...DEFAULT_ERROR_RESPONSES,
  },
})

const getDatasetRoute = createRoute({
  method: 'get',
  path: '/v1/datasets/{ref}',
  responses: {
    200: jsonResponse(DatasetResponseSchema, 'Dataset manifest'),
    ...DEFAULT_ERROR_RESPONSES,
  },
})

const samplesRoute = createRoute({
  method: 'get',
  path: '/v1/datasets/{ref}/samples',
  request: {
    query: PaginationQuerySchema,
  },
  responses: {
    200: jsonResponse(SamplesPageOpenApiSchema, 'Paginated samples'),
    ...DEFAULT_ERROR_RESPONSES,
  },
})

// NDJSON streaming responses can't be expressed by `app.openapi()`: the
// `application/x-ndjson` media type makes @hono/zod-openapi infer the 200 body
// as `never`, so a streaming `Response` is not assignable to the handler type.
// We register the OpenAPI path for documentation and serve it with a plain
// (fully typed, cast-free) `app.get` handler below.
const exportRoute = createRoute({
  method: 'get',
  path: '/v1/datasets/{ref}/export',
  request: {
    query: ExportDatasetQuerySchema,
  },
  responses: {
    200: {
      description: 'Dataset export as NDJSON',
      content: {
        'application/x-ndjson': {
          schema: z.string(),
        },
      },
    },
    ...DEFAULT_ERROR_RESPONSES,
  },
})

export function registerDatasetRoutes(app: OpenAPIHono<ApiEnv>): void {
  app.openapi(ingestSamplesRoute, async (context) => {
    // The route schema is a surrogate (SampleSchema embeds the unrenderable
    // JSON-number lexeme), so validate the full body with the real schema here.
    const body = IngestSamplesRequestSchema.parse(await context.req.json())
    const dataset = await getWorkspace(context).addSamples(body.samples, {
      name: body.name,
      message: body.message,
    })

    return context.json(dataset.manifest, 200)
  })

  app.openapi(ingestJsonlRoute, async (context) => {
    const query = context.req.valid('query')
    const body = await context.req.parseBody()
    const file = body.file

    if (!isFile(file)) {
      throw new BadInputError('multipart field "file" is required')
    }

    const tempPath = join(tmpdir(), `databench-${randomUUID()}.jsonl`)

    try {
      await writeFile(tempPath, Buffer.from(await file.arrayBuffer()))
      const dataset = await getWorkspace(context).addJsonl(tempPath, {
        ...(query.name !== undefined ? { name: query.name } : {}),
        ...(query.kind !== undefined ? { kind: query.kind } : {}),
        source: query.source ?? defaultSource(file),
      })

      return context.json(dataset.manifest, 200)
    } finally {
      await rm(tempPath, { force: true })
    }
  })

  app.openapi(getDatasetRoute, async (context) => {
    const dataset = await getWorkspace(context).get(context.req.param('ref'))

    return context.json(dataset.manifest, 200)
  })

  app.openapi(samplesRoute, async (context) => {
    const { limit, offset } = context.req.valid('query')
    const dataset = await getWorkspace(context).get(context.req.param('ref'))
    const items = [...dataset.toSamples(offset, limit)]

    return context.json(
      {
        total: dataset.length,
        limit,
        offset,
        items: items.map(toJsonCompatible),
      },
      200,
    )
  })

  app.openAPIRegistry.registerPath(exportRoute)
  app.get(exportRoute.getRoutingPath(), async (context) => {
    const { fmt } = ExportDatasetQuerySchema.parse(context.req.query())
    const ref = context.req.param('ref')

    const { filename, lines } = await getWorkspace(context).exportJsonl(ref, fmt)

    return context.body(streamLines(lines), 200, {
      'content-type': 'application/x-ndjson; charset=utf-8',
      'content-disposition': `attachment; filename="${safeFilename(filename)}"`,
    })
  })
}

function isFile(value: unknown): value is File {
  return value instanceof File
}

function defaultSource(file: File): string | null {
  return file.name ? parse(file.name).name : null
}

function safeFilename(filename: string): string {
  return filename.replaceAll('"', '_')
}

function streamLines(lines: Iterable<string>): ReadableStream<Uint8Array> {
  const encoder = new TextEncoder()

  return new ReadableStream({
    start(controller) {
      for (const line of lines) {
        controller.enqueue(encoder.encode(line))
      }

      controller.close()
    },
  })
}
